Replace async by promise_test

xiuqijix opened the pull request:
Combine following files to MediaStream-GUM.https.html:
GUM-api.https.html
GUM-empty-option-param.https.html
GUM-impossible-constraint.https.html
GUM-optional-constraint.https.html
GUM-trivial-constraint.https.html
GUM-unknownkey-option-param.https.html

<!-- Reviewable:start -->

<!-- Reviewable:end -->

xiuqijix committed:
Replace async by promise_test

w3c-bots commented:
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/19800)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/291435337)

xiuqijix committed:
Fix lint errors

Ms2ger commented on mediacapture-streams/GUM-deny.https.html:
```diff
@@ -16,18 +16,16 @@ <h1 class="instructions">Description</h1>
+      .catch(error => {
+        assert_equals(error.name, "NotAllowedError", "NotAllowedError returned as expected");
+        assert_equals(error.constraintName, undefined, "constraintName attribute not set as expected");
+      });
```
Use `promise_rejects(t, "NotAllowedError", () => navigator.mediaDevices.getUserMedia({video: true}))`

kaixinjxq replied:
@Ms2ger , thanks for your comment, fixed.

xiuqijix committed:
Use promise_rejects

alvestrand commented on mediacapture-streams/MediaStream-GUM.https.html:
```diff
@@ -0,0 +1,59 @@
+<script src=/resources/testharness.js></script>
+<script src=/resources/testharnessreport.js></script>
+
+<p class="instructions">Use a test device with color camera(embedded or external).</p>
```
Where did the idea  that the camera has to be a color camera come from?
"Camera" should be enough.

kaixinjxq replied:
@alvestrand, thanks for your comment, fixed.

alvestrand commented on mediacapture-streams/MediaStream-GUM.https.html:
```diff
@@ -0,0 +1,59 @@
+
+<p class="instructions">Use a test device with color camera(embedded or external).</p>
+<p class="instructions">When prompted, accept to share your color stream.</p>
+
```
"your color stream" should be "camera".

kaixinjxq replied:
Fixed.

alvestrand commented on mediacapture-streams/MediaStream-GUM.https.html:
```diff
@@ -0,0 +1,59 @@
+promise_test(t => {
+  return promise_rejects(t, new TypeError(), navigator.mediaDevices.getUserMedia({doesnotexist: true}));
+}, "Tests that getUserMedia is rejected with a TypeError when used with an unknown constraint");
+
```
This description is wrong. It's not the constraint that's unknown, it's the type of device requested.

kaixinjxq replied:
Fixed.

kaixinjxq replied:
I will fix it this PR.

alvestrand commented on mediacapture-streams/MediaStream-GUM.https.html:
```diff
@@ -0,0 +1,59 @@
+  })
+  .catch(error => {
+    assert_equals(error.name, "OverconstrainedError", "An impossible constraint triggers a OverconstrainedError");
+    assert_equals(error.constraint, "width", "The name of the not satisfied error is given in error.constraint");
```
English nit: "the not satisfied error" => "the unsatisfied constraint"

kaixinjxq replied:
Fixed.

alvestrand reviewed (commented):
This seems like a good thing to do. The new tests are more readable.
A couple of comments on wordings that should be improved in the reformulated test conditions.

xiuqijix committed:
Fix the description issues

kaixinjxq commented:
@alvestrand, I fixed the description issues and solved the conflicts, PTAL.

kaixinjxq commented:
@alvestrand, friendly ping. :)

alvestrand commented:
Apologies, this got lost. @kaixinjxq do you wish to rebase or abandon?
We're trying to step up our game on wpt review, so I hope this kind of stall stops happening.

kaixinjxq commented:
I'd like to rebase and refresh it with `async/await`.

xiuqijix committed:
Rebase to master and use async/await

- Rebase to master
- refresh the code with async/await
- Combine GUM-invalid-facing-mode.https.html to MediaStream-GUM.https.html
- Delete redundant <html><head><body> elements
- Delete obsolete MediaStream inactive events in MediaStream-idl.https.html

kaixinjxq commented:
@alvestrand, I just submit a new commit, PTAL.

alvestrand commented on mediacapture-streams/MediaDevices-getUserMedia.https.html:
```diff
@@ -37,60 +35,57 @@ <h1 class="instructions">Description</h1>
+      assert_equals(found_device.kind, "videoinput");
+      assert_equals(found_device.groupId, devices[i].groupId);
+    } catch (e) {
+      assert_equals(e.name, "OverconstrainedError");
```
Are you supposed to hit this branch, or not?
If you sometimes hit it, sometimes not, please make a comment to that effect.
If you expect to hit it every time, please find a way to fail the test if you don't.

kaixinjxq replied:
This test traverses all available devices and part of them maybe overconstrained if it does not a video device, that's why it catches the error and make it as expected result. And it's the original logic, I just rewrote it with `async/await`.

kaixinjxq replied:
I will fix it in the next PR.

alvestrand commented on mediacapture-streams/MediaDevices-getUserMedia.https.html:
```diff
@@ -37,60 +35,57 @@ <h1 class="instructions">Description</h1>
+      assert_equals(found_device.kind, "audioinput");
+      assert_equals(found_device.groupId, devices[i].groupId);
+    } catch (e) {
+      assert_equals(e.name, "OverconstrainedError");
```
Same comment as above.

kaixinjxq replied:
I will fix it in the next PR.

alvestrand reviewed (commented):
I have given this a cursory review and think it looks good. Want @jan-ivar to have an opinion too.

jan-ivar commented on mediacapture-streams/MediaStream-gettrackid.https.html:
```diff
@@ -1,32 +1,22 @@
+  const track = stream.getVideoTracks()[0];
+  t.add_cleanup(() => track.stop());
+  assert_equals(track, stream.getTrackById(track.id), "getTrackById returns track of given id");
+  assert_equals(stream.getTrackById(track.id + "foo"), null, "getTrackById of inexistant id returns null");
```
inexistent

kaixinjxq replied:
I will fix it this PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-MediaElement-disabled-video-is-black.https.html:
```diff
@@ -1,53 +1,45 @@
+  const stream = await navigator.mediaDevices.getUserMedia({video: true});
+  const [track] = stream.getVideoTracks();
+  t.add_cleanup(() => track.stop());
+  if (track.enabled) {
```
redundant `if`

kaixinjxq replied:
I will fix it this PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-applyConstraints.https.html:
```diff
@@ -8,51 +8,46 @@
+    const stream = await navigator.mediaDevices.getUserMedia({ video: true });
+    const [track] = stream.getVideoTracks();
+    t.add_cleanup(() => track.stop());
+    const groupId = track.getSettings().groupId;
```
Maybe assert that `groupId` is not `undefined`?

kaixinjxq replied:
I will fix it this PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-applyConstraints.https.html:
```diff
@@ -8,51 +8,46 @@
+    const devices = await navigator.mediaDevices.enumerateDevices();
+    const anotherDevice = devices.find(device => {
+      return device.kind == "videoinput" && device.groupId != groupId;
+    });
```
We could use destructoring:
```js
const settings = track.getSettings();
const anotherDevice = devices.find(({kind, groupId}) =>
    kind == "videoinput" && groupId != settings.groupId);
```

kaixinjxq replied:
I will fix it this PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-applyConstraints.https.html:
```diff
@@ -8,51 +8,46 @@
+    const anotherDevice = devices.find(device => {
+      return device.kind == "videoinput" && device.groupId != groupId;
+    });
+    if (anotherDevice !== undefined) {
```
Shouldn't we assert instead of skipping test here? Seems rare for a test system to have one device but not two (audio and video)...

kaixinjxq replied:
I will fix it in the next PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-getSettings.https.html:
```diff
@@ -8,72 +8,78 @@
+    const settings2 = stream2.getVideoTracks()[0].getSettings();
+    assert_equals(settings1.deviceId, settings2.deviceId);
+    assert_equals(settings1.width / 2, settings2.width);
+  }, 'A device can be opened twice with different resolutions');
```
There is no spec-support for this test. There's no guarantee the width and height will be different here, let alone exactly half. A browser that does not support downscaling is still spec-compliant.

kaixinjxq replied:
So, can we delete this test? I will remove it in the next PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-getSettings.https.html:
```diff
@@ -8,72 +8,78 @@
+    t.add_cleanup(() => tracks.forEach(track => track.stop()));
+    for (const device of devices) {
+      if (device.kind == "audiooutput")
+        continue;
```
Nit: maybe add `{` `}`, or put on a single line?

kaixinjxq replied:
I will fix it this PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-getSettings.https.html:
```diff
@@ -8,72 +8,78 @@
+      const device_id_constraint = {deviceId: {exact: device.deviceId}};
+      const constraints = device.kind == "audioinput"
+        ? {audio: device_id_constraint}
+        : {video: device_id_constraint};
```
This is going to do the wrong thing for `audiooutput`.

kaixinjxq replied:
The `audiooutput` has already been ruled out in L70~L71.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-getSettings.https.html:
```diff
@@ -8,72 +8,78 @@
+        : {video: device_id_constraint};
+
+      const stream = await navigator.mediaDevices.getUserMedia(constraints);
+      tracks.push(...stream.getTracks());
```
Again I would `stop()` tracks in the for-loop here for better odds of success.

kaixinjxq replied:
I will fix it this PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-init.https.html:
```diff
@@ -1,37 +1,32 @@
+  const stream = await navigator.mediaDevices.getUserMedia({video: true});
+  const videoTracks = stream.getVideoTracks();
+  assert_equals(videoTracks.length, 1, "There is exactly one video track in the media stream");
+  track = videoTracks[0];
```
```js
const track = videoTracks[0];
```

kaixinjxq replied:
I will fix it this PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrack-init.https.html:
```diff
@@ -1,37 +1,32 @@
+  assert_equals(track.readyState, "live", "The track object is in live state");
+  assert_equals(track.kind, "video", "The track object is of video kind");
+  // Not clear that this is required by the spec,
+  // see https://www.w3.org/Bugs/Public/show_bug.cgi?id=22212
```
Resolved. [Spec](https://w3c.github.io/mediacapture-main/getusermedia.html#life-cycle-and-media-flow) says: *"For a newly created MediaStreamTrack object, the following applies. The track is always enabled unless stated otherwise"*

kaixinjxq replied:
I will fix it this PR.

jan-ivar commented on mediacapture-streams/MediaStreamTrackEvent-constructor.https.html:
```diff
@@ -4,39 +4,41 @@
+  const track = dest.stream.getTracks()[0];
   assert_true(track instanceof MediaStreamTrack);
-  var event = new MediaStreamTrackEvent("type", { track: track });
+  const event = new MediaStreamTrackEvent("type", { track: track });
```
Tip:
```js
const event = new MediaStreamTrackEvent("type", { track });
```

kaixinjxq replied:
I will fix it this PR.

jan-ivar reviewed (changes requested):
Nicely done! Sorry for all the comments. Haven't looked at these tests in detail before.

There are some tests in here that are wrong, so we should either fix or file a follow-up, I think.

xiuqijix committed:
Address jan-ivar's comments

kaixinjxq commented on mediacapture-streams/MediaDevices-getUserMedia.https.html:
```diff
@@ -37,60 +35,57 @@ <h1 class="instructions">Description</h1>
+promise_test(async t => {
+  const devices = await navigator.mediaDevices.enumerateDevices();
+  const tracks = [];
+  t.add_cleanup(() => tracks.forEach(track => track.stop()));
```
OK, I will fix it this PR.

kaixinjxq commented on mediacapture-streams/MediaDevices-getUserMedia.https.html:
```diff
@@ -37,60 +35,57 @@ <h1 class="instructions">Description</h1>
+  const devices = await navigator.mediaDevices.enumerateDevices();
+  const tracks = [];
+  t.add_cleanup(() => tracks.forEach(track => track.stop()));
+  for (const i in devices) {
```
I will fix it this PR.

kaixinjxq commented on mediacapture-streams/MediaDevices-getUserMedia.https.html:
```diff
@@ -37,60 +35,57 @@ <h1 class="instructions">Description</h1>
+      );
+      tracks.push(...stream.getTracks());
+      const found_device = devices.find(element => {
+        return element.deviceId == stream.getTracks()[0].getSettings().deviceId;
```
I will fix it this PR.

kaixinjxq commented on mediacapture-streams/MediaDevices-getUserMedia.https.html:
```diff
@@ -37,60 +35,57 @@ <h1 class="instructions">Description</h1>
+      const found_device = devices.find(element => {
+        return element.deviceId == stream.getTracks()[0].getSettings().deviceId;
+      });
+      assert_true(undefined !== found_device);
```
I will fix it this PR.

kaixinjxq commented on mediacapture-streams/MediaStream-MediaElement-preload-none-manual.https.html:
```diff
@@ -1,58 +1,41 @@
+  const aud = document.querySelector("audio");
+  const stream = await navigator.mediaDevices.getUserMedia({audio:true});
+  t.add_cleanup(() => stream.getAudioTracks()[0].stop());
+  testPreloadNone(t, aud, () => { aud.srcObject = stream; });
```
I will fix it this PR.

kaixinjxq commented on mediacapture-streams/MediaStream-MediaElement-preload-none-manual.https.html:
```diff
@@ -1,58 +1,41 @@
+  const vid = document.querySelector("video");
+  const stream = await navigator.mediaDevices.getUserMedia({video:true});
+  t.add_cleanup(() => stream.getVideoTracks()[0].stop());
+  testPreloadNone(t, vid, () => { vid.srcObject = stream; });
```
I will fix it this PR.

kaixinjxq commented on mediacapture-streams/MediaStream-add-audio-track.https.html:
```diff
@@ -1,53 +1,40 @@
+  assert_equals(video.getAudioTracks().length, 1, "video mediastream still has one audio track"); // If track is already in stream's track set, then abort these steps.
+
+  audio.onaddtrack = t.step_func(() =>
+    assert_unreached("onaddtrack is not fired when the script directly modified the track of a mediastream"));
```
Sorry, I didn't get the point, actually `audio.onaddtrack` is already above the `audio.addTrack`.

kaixinjxq commented on mediacapture-streams/MediaStream-add-audio-track.https.html:
```diff
@@ -1,53 +1,40 @@
+
+  const tracks = audio.getTracks();
+  t.add_cleanup(() => tracks.forEach(track => track.stop()));
+  tracks.push(...video.getTracks());
```
I will fix it this PR.

kaixinjxq commented on mediacapture-streams/MediaStream-default-feature-policy.https.html:
```diff
@@ -1,62 +1,56 @@
+// mic/camera has been explicitly allowed by feature policy.
+function promise_factory(allowed_features) {
+  return new Promise((resolve, reject) => {
+    navigator.mediaDevices.getUserMedia({video: true, audio: true}).then(
```
I will fix it in the next PR.

kaixinjxq commented on mediacapture-streams/MediaStreamTrack-init.https.html:
```diff
@@ -1,37 +1,32 @@
 <!doctype html>
-<html>
-<head>
 <title>getUserMedia({video:true}) creates a stream with a properly initialized video track</title>
```
I will fix it this PR.

kaixinjxq commented:
@jan-ivar, many thanks for your review! I just submit a new commit to address your comments, PTAL. Some suggestions need to modify the logic of the source code, I would like to fix these in a later PR.

kaixinjxq commented:
This PR is blocked on the required "Travis CI - Pull Request" check after #14499. In order to trigger it, I will close and reopen this PR.
